### ISSUE TITLE
how to training a new language

### ISSUE THREAD

**Issue by MonolithFoundation** [lucasnewman/nanospeech#3]
hi, what further need to do for support a new language training?

**Comment by lucasnewman** [comment]
The primary thing you need is a dataset with speech and transcriptions (or the ability to generate transcriptions with e.g. whisper). Do you have that available?

If you can get that into a form where you can cycle it in a dataloader, you should be ready to train on a GPU -- if you report back I can help you get it going!

**Comment by MonolithFoundation** [comment]
Yes. am not sure if training on Chinese dataset only with multiple speaker is OK or not, would that make pretrained forget English tts ability?

**Comment by lucasnewman** [comment]
> Yes. am not sure if training on Chinese dataset only with multiple speaker is OK or not, would that make pretrained forget English tts ability?

I think you should just target training your own vs using the pretrained model, as the vocabulary wouldn't support both languages. If you assemble a mix of Chinese and English, multi-speaker or otherwise, it should work.

If you need an English dataset, you can use [LibriTTS-R](https://www.openslr.org/141/) — it's small but sufficient to get basic text alignment going. Increasing the scale of your data will help if you're looking for higher quality results, of course.

One option if you're using `datasets` (the HF library) would be to just take both a Chinese and English dataset and [concatenate them](https://huggingface.co/docs/datasets/v3.2.0/en/package_reference/main_classes#datasets.concatenate_datasets) at training time, and then cycle that in a single dataloader.

**Comment by lucasnewman** [comment]
> Very cool project! Can you give a bit of guidance on training? How many steps are expected (roughly)? Is the duration model supposed to be trained prior to the main model training? The current `train_nanospeech.py` example doesn't appear to load a pre-trained duration model.

It depends on your level of compute / dataset size / batch size, but I would target at least 1 million steps to get proper alignment. For what it's worth, I did this on a single GH200 (H100 + arm64 cpu), which I got for $1.49/hr from [Lambda](https://lambdalabs.com). The total cost to run 1M steps was $156, and took about 104 hours total.

You can train the duration model last, because during training you're using a random mask prediction objective, and you'll know the duration of the original sample, so you don't have to do any guessing. At inference time, you can just hand-pick durations for testing (it's flexible around slightly different durations -- the speaking rate will speed up or slow down), and then once you're happy with the speech generation model, you can train a duration model for about 10% of the compute as the main model.

I'll see if I can add some kind of annotated tutorial that makes it clear how to go from dataset -> WebDataset shards -> training both models that makes it as clear as possible.

**Comment by lucasnewman** [comment]
I put up a quick tutorial [here](https://github.com/lucasnewman/nanospeech/tree/main/tutorial) on training a model from scratch. It's not finished, but it should be enough to get started training -- let me know if you run into issues!

**Comment by roedoejet** [comment]
This is awesome @lucasnewman - thank you! In our case, compute isn't really an issue, but we are somewhat limited by dataset size. Most datasets we have are between 1 and 10 hours per language. The tutorial is very nice thank you. We'll try this out and post if we have any issues.
